From 52eeeb713725bed8db2d0ba098e105561e8f0bdd Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:02:43 -0400
Subject: [usb] ehci: add check for wakeup/suspend race

Message-id: <1410887463-20674-92-git-send-email-dzickus@redhat.com>
Patchwork-id: 95121
O-Subject: [RHEL7 PATCH 091/591] USB: EHCI: add check for wakeup/suspend race
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit b8efdafb4005a79ab2305f8af31025d49292d4cf
Author: Alan Stern <stern@rowland.harvard.edu>
Date:   Fri Oct 18 11:18:21 2013 -0400

    USB: EHCI: add check for wakeup/suspend race

    hcd-pci.c in usbcore contains a check for wakeup requests racing with
    controller suspend.  This check is going to be moved out of usbcore
    and into the individual controller drivers, where it can apply to all
    platforms, not just PCI.

    This patch adds the check to ehci-hcd.

    Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/host/ehci-hcd.c b/drivers/usb/host/ehci-hcd.c
index 606e18c..e3afac7 100644
--- a/drivers/usb/host/ehci-hcd.c
+++ b/drivers/usb/host/ehci-hcd.c
@@ -1095,6 +1095,14 @@ int ehci_suspend(struct usb_hcd *hcd, bool do_wakeup)
  clear_bit(HCD_FLAG_HW_ACCESSIBLE, &hcd->flags);
  spin_unlock_irq(&ehci->lock);
 
+ synchronize_irq(hcd->irq);
+
+ /* Check for race with a wakeup request */
+ if (do_wakeup && HCD_WAKEUP_PENDING(hcd)) {
+  ehci_resume(hcd, false);
+  return -EBUSY;
+ }
+
  return 0;
 }
 EXPORT_SYMBOL_GPL(ehci_suspend);
-- 
1.7.1